mqtt/mqtt-broker/src/session.rs

```diff
@@ -910,6 +909,99 @@ impl Default for PacketIdentifiers {
     }
 }
 
+/// `BoundedQueue` is a queue of publications with bounds by count and total payload size in bytes.
+///
+/// Packets will be queued until either `max_len` (max number of publications)
+/// or `max_size` (max total payload size of publications)
+/// is reached, and then `when_full` strategy is applied.
+///
+/// None for `max_len` or `max_size` means "unbounded".
+#[derive(Clone, Debug, PartialEq)]
+struct BoundedQueue {
+    inner: VecDeque<proto::Publication>,
+    max_len: Option<NonZeroUsize>,
+    max_size: Option<NonZeroU64>,
+    when_full: QueueFullAction,
+    current_size: u64,
+}
+
+impl BoundedQueue {
+    pub fn new(
+        max_len: Option<NonZeroUsize>,
+        max_size: Option<NonZeroU64>,
+        when_full: QueueFullAction,
+    ) -> Self {
+        BoundedQueue {
+            inner: VecDeque::new(),
+            max_len,
+            max_size,
+            when_full,
+            current_size: 0,
+        }
+    }
+
+    pub fn into_inner(self) -> VecDeque<proto::Publication> {
+        self.inner
+    }
+
+    pub fn dequeue(&mut self) -> Option<proto::Publication> {
+        match self.inner.pop_front() {
+            Some(publication) => {
+                self.current_size -= publication.payload.len() as u64;
+                Some(publication)
+            }
+            None => None,
+        }
+    }
+
+    pub fn enqueue(&mut self, publication: proto::Publication) {
+        if let Some(max_len) = self.max_len {
+            if self.inner.len() >= max_len.get() {
+                return self.handle_queue_limit(publication);
+            }
+        }
+
+        if let Some(max_size) = self.max_size {
+            let pub_len = publication.payload.len() as u64;
+            if self.current_size + pub_len > max_size.get() {
+                return self.handle_queue_limit(publication);
+            }
+        }
+
+        self.current_size += publication.payload.len() as u64;
+        self.inner.push_back(publication);
+    }
+
+    #[cfg(test)]
```

Comment:
nit: should we make it available for everything?

Reply:
So far it just used in tests, otherwise give dead code warning. We can remove `cfg` when we need to use this in prod code.